src/sshfs_mount/sshfs_mount_handler.cpp

```diff
@@ -60,229 +48,161 @@ void start_and_block_until(mp::Process* process, Signal signal, std::function<bo
-            mpl::log(mpl::Level::warning, category,
-                     fmt::format("Failed to install \'multipass-sshfs\', error message: \'{}\'",
-                                 mp::utils::trim_end(error_msg)));
-            throw mp::SSHFSMissingError();
-        }
-    }
-    catch (const mp::ExitlessSSHProcessException&)
+void install_sshfs_for(const std::string& name, mp::SSHSession& session, const std::chrono::milliseconds& timeout)
+try
+{
+    mpl::log(mpl::Level::info, category, fmt::format("Installing the multipass-sshfs snap in '{}'", name));
+    auto proc = session.exec("sudo snap install multipass-sshfs");
+    if (proc.exit_code(timeout) != 0)
     {
-        mpl::log(mpl::Level::info, category, fmt::format("Timeout while installing 'sshfs' in '{}'", name));
+        auto error_msg = proc.read_std_error();
+        mpl::log(mpl::Level::warning, category,
+                 fmt::format("Failed to install 'multipass-sshfs': {}", mp::utils::trim_end(error_msg)));
         throw mp::SSHFSMissingError();
     }
 }
-} // namespace
-
-mp::SSHFSMountHandler::SSHFSMountHandler(const SSHKeyProvider& key_provider) : MountHandler(key_provider)
+catch (const mp::ExitlessSSHProcessException&)
 {
+    mpl::log(mpl::Level::info, category, fmt::format("Timeout while installing 'multipass-sshfs' in '{}'", name));
+    throw mp::SSHFSMissingError();
 }
+} // namespace
 
-void mp::SSHFSMountHandler::init_mount(VirtualMachine* vm, const std::string& target_path, const VMMount& vm_mount)
+namespace multipass
+{
+SSHFSMountHandler::SSHFSMountHandler(VirtualMachine* vm, const SSHKeyProvider* ssh_key_provider, std::string target,
+                                     const VMMount& mount)
+    : MountHandler{vm, ssh_key_provider, target, mount}
 {
-    if (!MP_FILEOPS.exists(QDir{QString::fromStdString(vm_mount.source_path)}))
-    {
-        mount_processes[vm->vm_name].erase(target_path);
-        throw std::runtime_error(fmt::format("Mount path \"{}\" does not exist.", vm_mount.source_path));
-    }
-
     mpl::log(mpl::Level::info, category,
-             fmt::format("initializing mount {} => {} in {}", vm_mount.source_path, target_path, vm->vm_name));
+             fmt::format("initializing mount {} => {} in '{}'", mount.source_path, target, vm->vm_name));
 
-    // Putting this here is kind of contrived, but passing all the same arguments to start_mount() is
-    // redundant.
-    mp::SSHFSServerConfig config;
     config.port = vm->ssh_port();
     config.username = vm->ssh_username();
     config.instance = vm->vm_name;
-    config.target_path = target_path;
-    config.source_path = vm_mount.source_path;
-    config.uid_mappings = vm_mount.uid_mappings;
-    config.gid_mappings = vm_mount.gid_mappings;
+    config.target_path = target;
+    config.source_path = mount.source_path;
+    config.uid_mappings = mount.uid_mappings;
+    config.gid_mappings = mount.gid_mappings;
     config.private_key = ssh_key_provider->private_key_as_base64();
```

Comment:
I think that would be a little ugly without the designated initializers of C++20. The first field of `SSHFSServerConfig` is `host`, but that one gets assigned only in `start`, so in the initializer list I would probably have to have something like
```cpp
: MountHandler{vm, ssh_key_provider, target, mount},
  config{"",
         vm->ssh_port(),
         vm->ssh_username(),
         vm->vm_name,
         ssh_key_provider->private_key_as_base64(),
         mount.source_path,
         target,
         mount.uid_mappings,
         mount.gid_mappings}
```
Which I think is kind of error prone, in case something changes in `SSHFSServerConfig`, like the order of the fields.

Reply:
Yeah, I understand your point, and yes, naming parameters will be really nice. But the risk of wrong parameter ordering affects any function with multiple parameters of the same type. Ideally, one would create different types for the different strings, to make it safer. But I think it is not worth the trouble. Granted, with regular functions you don't have an alternative like the one you use here, but I am not quite convinced it is better. It is also error prone, in that it is easy to leave stuff uninitialized if new fields are added. And being forced to pass that empty string makes it clearer that it is not there yet. You might even pass "not yet defined" or something.

But OK, all in all I recognize this is splitting hairs and there is a margin of subjectivity. It's your code and you have a good justification, so leave it as you prefer. And feel free to argue and convince me anyway if you want :)